src/Data/List.purs

```diff
@@ -257,6 +253,18 @@ uncons :: forall a. List a -> Maybe { head :: a, tail :: List a }
 uncons Nil = Nothing
 uncons (x : xs) = Just { head: x, tail: xs }
 
+-- | Break a list into its last element, and the preceding elements,
+-- | or `Nothing` if the list is empty.
+-- |
+-- | Running time: `O(n)`
+unsnoc :: forall a. List a -> Maybe { init :: List a, last :: a }
+unsnoc lst = unsnocHelper lst Nil <#> \h -> { init: reverse h.revInit, last: h.last }
+
+unsnocHelper :: forall a. List a -> List a -> Maybe { revInit :: List a, last :: a }
```

Comment:
Sure! The main reason I split it off was concerns about performance – I figured that redefining `go` on every call to `unsnoc` was wasteful. But I don't know much about that, so I'll take your advice.